File: Difficulty: Medium/Longest subarray with Atmost two distinct integers/longest-subarray-with-atmost-two-distinct-integers.java
class Solution {
    public int totalElements(int[] arr) {
        int n = arr.length;
        Map<Integer, Integer> map = new HashMap<>();

        int i      = 0;       // left boundary of window
        int j      = 0;       // right boundary of window
        int result = 0;       // best window size seen

        while (j < n) {
            // include arr[j] into the window
            map.put(arr[j], map.getOrDefault(arr[j], 0) + 1);

            // if more than two distinct numbers, shrink from left
            while (map.size() > 2) {
                int count = map.get(arr[i]) - 1;
                if (count == 0) {
                    map.remove(arr[i]);     // remove key if freq is zero
                } else {
                    map.put(arr[i], count); // else update freq
                }
                i++;                        // move left boundary right
            }

            // window [i..j] is valid, update result
            result = Math.max(result, j - i + 1);
            j++;                           // expand window to the right
        }

        return result;
    }
}

File: Difficulty: Medium/Longest subarray with Atmost two distinct integers/longest-subarray-with-atmost-two-distinct-integers.cpp
class Solution {
  public:
    int totalElements(vector<int>& arr) {
        int n = arr.size();
        unordered_map<int, int> mp;

        int i      = 0;          // left boundary of window
        int j      = 0;          // right boundary of window
        int result = 0;          // best window size seen

        while (j < n) {
            // include arr[j] into the window
            mp[arr[j]]++;

            // if more than two distinct numbers, shrink from left
            while (mp.size() > 2) {
                mp[arr[i]]--;            // decrement freq of arr[i]
                if (mp[arr[i]] == 0) {   // if freq is zero, remove key
                    mp.erase(arr[i]);
                }
                i++;                     // move left boundary right
            }

            // window [i..j] is valid, update result
            result = max(result, j - i + 1);
            j++;                        // expand window to the right
        }

        return result;
    }
};